threescale/src/structs.rs

```diff
@@ -45,11 +45,168 @@ pub struct UsageReport {
+        Self(a.to_string())
+    }
+}
+
+impl From<&str> for AppId {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+impl From<&str> for AppKey {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+#[repr(transparent)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ServiceToken(String);
+#[repr(transparent)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ServiceId(String);
+
+impl AsRef<str> for ServiceToken {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl AsRef<str> for ServiceId {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl From<&str> for ServiceToken {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+impl From<&str> for ServiceId {
+    fn from(a: &str) -> Self {
+        Self(a.to_string())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CacheKey(ServiceId, AppIdentifier);
+
+impl<'a> CacheKey {
+    pub fn as_string(&self) -> String {
+        format!("{}_{}", self.0 .0, self.1.as_string())
+    }
+
+    pub fn default() -> CacheKey {
+        CacheKey(
+            ServiceId(String::new()),
+            AppIdentifier::UserKey(UserKey(String::new())),
+        )
+    }
+
+    pub fn service_id(&'a self) -> &'a ServiceId {
+        &self.0
+    }
+
+    pub fn app_id(&'a self) -> &'a AppIdentifier {
+        &self.1
+    }
+
+    pub fn from(a: &ServiceId, b: &AppIdentifier) -> CacheKey {
+        CacheKey {
+            0: a.clone(),
+            1: b.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum AppIdentifier {
+    AppId(AppId, Option<AppKey>),
+    UserKey(UserKey),
+}
+
+impl From<AppId> for AppIdentifier {
+    fn from(a: AppId) -> Self {
+        AppIdentifier::AppId(a, None)
+    }
+}
+
+impl From<(AppId, AppKey)> for AppIdentifier {
+    fn from(a: (AppId, AppKey)) -> Self {
+        AppIdentifier::AppId(a.0, Some(a.1))
+    }
+}
+
+impl From<UserKey> for AppIdentifier {
+    fn from(u: UserKey) -> Self {
+        AppIdentifier::UserKey(u)
+    }
+}
+
+impl AppIdentifier {
+    // This cannot return a reference because app_id:app_key needs to be generated
+    pub fn as_string(&self) -> String {
+        match self {
+            AppIdentifier::AppId(AppId(id), key) => {
+                let mut res: String = id.clone();
+                if let Some(AppKey(app_key)) = key {
+                    res.push(':');
+                    res.push_str(app_key.as_str());
+                }
+                res
+            }
+            AppIdentifier::UserKey(UserKey(user_key)) => user_key.clone(),
+        }
+    }
+
+    pub fn appid_from_str(s: &str) -> AppIdentifier {
+        let v: Vec<&str> = s.split(':').collect();
+        if v.len() == 1 {
+            return AppIdentifier::AppId(AppId(v[0].to_owned()), Some(AppKey(v[1].to_owned())));
+        }
+        AppIdentifier::AppId(AppId(v[0].to_owned()), None)
+    }
+}
```

Comment:
When you generate a `Self` (ie. AppIdentifier) in an impl block, consider whether you would be best served by implementing `From`, `TryFrom`, or `FromStr`, which have the benefits of working in more scenarios because they are well known traits for which people provide more ergonomic helpers and generic conversions, such as being able to call `"app_id"::parse::<AppIdenfitier>()` if you `impl FromStr`, and being able to specify bounds like `A: Into<AppIdentifier>` if you impl `From<...> for AppIdentifier` (as any `From<Src> for Target` automatically impls `Into<Target> for Src`, though not the other way around).